### ISSUE TITLE
UX issues with qrexec service confirmation dialogs - missing target default value

### ISSUE THREAD

**Issue by na--** [QubesOS/qubes-issues#3251]
#### Qubes OS version:

R4.0 RC2

#### Affected TemplateVMs: 

all (dom0 issue)

---

### Steps to reproduce the behavior:

Try to execute a simple Qubes RPC call, for example try to copy a file from one VM to another with `qvm-copy-to-vm`, plug in an external keyboard/mouse, use [VMAuth](https://www.qubes-os.org/doc/vm-sudo/#replacing-password-less-root-access-with-dom0-user-prompt) (yeah, I know...:) ), sign a git commit with split gpg, etc. - most RPC actions with an `ask` policy.

### Expected behavior:

A simple dialog opens and with minimum friction the user chooses whether to allow the requested action or not.

### Actual behavior:

The "Operation execution" dialog is missing a default `target` value and it always has to be chosen from the dropdown. This leads to significantly more friction (instead of "escape or enter" / "yes or no" quick choice by the user), first the dropdown value has to be populated.

![screenshot_2017-10-28_18-45-27](https://user-images.githubusercontent.com/3336773/32136284-7cbcc296-bc13-11e7-8e50-5633b436fb71.png)

![screenshot_2017-10-28_18-42-14](https://user-images.githubusercontent.com/3336773/32136349-d19adf4a-bc14-11e7-8bcf-b932540098da.png)

### General notes:

That behavior is ok for calls like `qvm-copy` that do not specify the destination VM initially, but it's annoying for:
- actions like `VMAuth` and `InputMouse`/`InputKeyboard` have only one possible target in the dropdown and it should be pre-populated or missing altogether.
- actions like `qvm-move-to-vm`, `qvm-copy-to-vm` for which I have to write the destination VM twice - once in the command itself and once in the dom0 dialog
- automation related stuff like signing git commits with `split-gpg` - what's the point of specifying `/rw/config/gpg-split-domain` if I have to enter it again every time anyway

---

#### Related issues:

None that I could find

**Comment by andrewdavidwong** [comment]
I believe this is intentional, as explained here:
https://www.qubes-os.org/news/2017/10/03/core3/

@marmarek, what do you think?

**Comment by marmarek** [comment]
Yes, this is intentional. If for some particular service it is desired to have some default value there, the qrexec policy needs to be adjusted for that, adding `default_target=` option. For example:
```
$anyvm sys-usb ask,default_target=sys-usb
```

As for `qvm-copy-to-qvm`, in default policy the target parameter is redundant. But it is left there for compatibility reasons. There are friendly wrappers without this argument: `qvm-copy`, `qvm-move` - there you provide target VM only once - in confirmation window.

So, I think documentation for some services needs to be adjusted (qubes.VMAuth, split gpg etc), to include `default_target=` if desired.

**Comment by na--** [comment]
Thanks, that solves a lot of my issues. The [qrexec](https://www.qubes-os.org/doc/qrexec3) documentation also needs to be adjusted, there is no mention of `default_target=` there.

**Comment by qubesuser** [comment]
I think the solution could be to have qvm-copy-to-vm and qvm-move-to-vm print a warning that they are deprecated and that qvm-copy and qvm-move should be used instead.

For instance, I just learned that by reading this bug.

**Comment by qubesuser** [comment]
And ideally some help text in the confirmation dialog that says "if you want to specify a target in the VM issuing the request, add a specific allow rule in the Qubes RPC policy for that specific source and target VM combination before the ask rule".

**Comment by marmarek** [comment]
That would encourage such configuration, which not what we want to achieve. In fact this change is especially to take out control over target domain from source domain where possible/sensible. But including such cases in documentation (where we have more space, also for recommendations) is a good idea.

**Comment by marmarek** [comment]
Added `default_target=` to documentation: https://github.com/QubesOS/qubes-doc/commit/1fb4b57b236b4558a9557cdc653854fbc9ce2a46

**Comment by andrewdavidwong** [comment]
Reopening for two reasons:

1. `qvm-copy` and `qvm-move` are missing from https://www.qubes-os.org/doc/tools/4.0/domU/.
2. `qvm-copy` appears to have an incorrect usage message, as reported in https://github.com/QubesOS/qubes-issues/issues/4207#issuecomment-412361138:
```
$ qvm-copy
usage: /usr/bin/qvm-copy-to-vm [--without-progress] dest_vmname file [file]+
$ qvm-copy-to-vm
usage: /usr/bin/qvm-copy-to-vm [--without-progress] dest_vmname file [file]+
```

**Comment by constantoverride** [comment]
This [doc](https://www.qubes-os.org/doc/tips-and-tricks/) is outdated because the confirmation(for `qvm-open-in-vm untrusted https://duckduckgo.com`) requires **re-entering** the target VM name(that is: `untrusted`) in the dialog that appears (and you can even override it by using a different VM name), so it's no longer true that 
> it will open duckduckgo.com in the `untrusted` AppVM (after you confirmed the request).

thus the `.desktop` file is kinda useless unless you already know what `APPVMNAME` was used inside it, for the purpose of re-entering its name every time an url is opened.

![askedscreenshot_2018-09-15_03-12-10](https://user-images.githubusercontent.com/40925413/45581154-8d9a3200-b899-11e8-9fe5-68a251f8dbeb.png)

**Comment by marmarek** [comment]
Indeed, it is slightly different on R4.0. For this to work, you need to add following line to `/etc/qubes-rpc/policy/qubes.OpenURL` (above default rules):

    SOURCE_APPVM untrusted ask,default_target=untrusted

(replace SOURCE_APPVM with actual VM name, or `@anyvm` if you want to apply it all VMs)

Alternatively, you can use `@default` as target VM in .desktop file and control default target solely from the policy using rule like this:

    SOURCE_APPVM @default ask,default_target=untrusted

The latter generally should be preferable, as the source VM doesn't learn about other VM names, but will not work if you want multiple .desktop file for opening links in different VMs.

**Comment by ben-grande** [comment]
> Yes, this is intentional. If for some particular service it is desired to have some default value there, the qrexec policy needs to be adjusted for that, adding `default_target=` option. For example:
> 
> ```
> $anyvm sys-usb ask,default_target=sys-usb
> ```
> 
> As for `qvm-copy-to-qvm`, in default policy the target parameter is redundant. But it is left there for compatibility reasons. There are friendly wrappers without this argument: `qvm-copy`, `qvm-move` - there you provide target VM only once - in confirmation window.
> 
> So, I think documentation for some services needs to be adjusted (qubes.VMAuth, split gpg etc), to include `default_target=` if desired.

It is [documented for split-gpg](https://doc.qubes-os.org/en/r4.3/user/security-in-qubes/split-gpg.html#advanced-configuration) and the Qrexec docs explains [default_target](https://doc.qubes-os.org/en/r4.3/developer/services/qrexec.html#specifying-vms-tags-types-targets-etc). The [VMAuth docs](https://doc.qubes-os.org/en/r4.3/user/security-in-qubes/vm-sudo.html#replacing-passwordless-root-access-with-dom0-user-prompt) are linked to the forum, where it contains `default_target` also.

---


> Reopening for three reasons:
> 
>     1. We need a similar solution for `qvm-open-in-vm`, as reported in

As show by Marek in the message above, a solution exists

>     3. `qvm-copy` and `qvm-move` are missing from https://www.qubes-os.org/doc/tools/4.0/domU/.
They are present now: https://doc.qubes-os.org/en/latest/user/reference/tools.html#domu.

---

Is there anything keeping this issue open?

**Comment by andrewdavidwong** [comment]
Closing as completed. If anyone believes this issue is not yet completed, or if anyone is still affected by this issue, please leave a comment saying so, and we'll be happy to reopen it. Thank you.
